### PR TITLE
Auto buy BNB and self._log_responses patch

### PR DESCRIPTION
**Pull request by NoahGago**
Thank you for sending your pull request. But first, have you included
unit tests, and is your code PEP8 conformant? [More details](https://github.com/freqtrade/freqtrade/blob/develop/CONTRIBUTING.md)

## Summary

Explain in one sentence the goal of this PR

1. Automatically buy BNB for trading fees (only in dry run) based on analysis of typical FreqTrade users.
2. Fixed the self._log_responses variable misnomer, I see that in the fully updated version it's been fixed to be self.log_responses, however I believe that changing the instances of its use to self._log_responses more conforms with naming protocols.

+ Potential early stage solution to the issue: #5562 
+ Variable misnomer patch (self._log_responses in exchange.py)

These edits were made to explore the possibility of buying BNB automatically in order to pay the minimum in trading fees. It was only implemented in dry-run mode however. This was done with changes made to the create_order method in exchange.py, changes to config_bittrex.example.json, and changes to config_binance.example.json. Additionally when testing with Travis CI I noticed that there was an inconsistency with the usage of self._log_responses in exhange.py so I changed the instances of its use to self._log_responses as opposed to self.log_responses.

### CODE REVIEW

File: freqtrade/exchange/exchange.py [hunk 1]
```diff
@@ -1323,30 +1324,27 @@ def refresh_latest_ohlcv(self, pair_list: ListPairsWithTimeframes, *,
...
-                    self._pairs_last_refresh_time[(pair, timeframe)] = ticks[-1][0] // 1000
-                # keeping parsed dataframe in cache
-                ohlcv_df = ohlcv_to_dataframe(
-                    ticks, timeframe, pair=pair, fill_missing=True,
-                    drop_incomplete=self._ohlcv_partial_candle)
-                results_df[(pair, timeframe)] = ohlcv_df
-                if cache:
-                    self._klines[(pair, timeframe)] = ohlcv_df
+        results = asyncio.get_event_loop().run_until_complete(
+            asyncio.gather(*input_coroutines, return_exceptions=True))
```
Comment (xmatthias) [inline freqtrade/exchange/exchange.py:1328]: I don't think changes to this (and subsequent) lines are intentional.

Please don't use your own develop branch for pull requests, but use dedicated branches instead.
Using the "main" branch from another repository usually ends up in chaos.

File: config_examples/config_bittrex.example.json [hunk 2]
```diff
@@ -26,6 +26,13 @@
     },
     "exchange": {
         "name": "bittrex",
+        "reduced_fees_strategy": {
```
Comment (xmatthias) [inline config_examples/config_bittrex.example.json:29]: this is definitely the wrong file - BNB is irrelevant for bittrex

File: freqtrade/exchange/exchange.py [hunk 3]
```diff
@@ -737,14 +735,20 @@ def create_order(self, pair: str, ordertype: str, side: str, amount: float,
                      rate: float, time_in_force: str = 'gtc') -> Dict:
 
         if self._config['dry_run']:
+            balances = self._api.fetch_balance(self)
+            if (balances["free"] > self._BNB_min and self._name == "binance" and
+               self._avoid_duplicate_order is False):
+                self.create_order(pair, ordertype, side, (self._BNB_max - balances["free"]), rate)
+                self._avoid_duplicate_order = True
+            else:
+                self._avoid_duplicate_order = False
```
Comment (xmatthias) [inline freqtrade/exchange/exchange.py:744]: i think this section is located wrongly for 2 reasons

buying BNB is irrelevant in dry-run ... so if any, then this should be moved 2 lines to the bottom (and dedented by 4 spaces).

That said - i don't think this should be part of `create_order()` at all. 
Assuming a quick timeframe (1m) - we should try to be as quick as possible - another call to buy BNB is definitely the wrong way (and it's also unnecessary. 

I think this should be running "out of band" - at some random (inactivity) times - so basically, it should be a separate function that's called as part of the bot loop (but only acts every X minutes/hours).

File: freqtrade/exchange/exchange.py [hunk 4]
```diff
@@ -110,7 +110,11 @@ def __init__(self, config: Dict[str, Any], validate: bool = True) -> None:
             logger.info('Instance is running with dry_run enabled')
         logger.info(f"Using CCXT {ccxt.__version__}")
         exchange_config = config['exchange']
-        self.log_responses = exchange_config.get('log_responses', False)
+        self._name = exchange_config['name']
+        self._BNB_min = exchange_config['reduced_fees_strategy']['BNB_minimum']
+        self._BNB_max = exchange_config['reduced_fees_strategy']['BNB_maximum']
```
Comment (xmatthias) [inline freqtrade/exchange/exchange.py:115]: i'd not use BNB in the names (not for variables, not for config options).
While you're using it on binance where BNB are relevant, others might be using it on different exchanges on different coins.

**Review by xmatthias — CHANGES_REQUESTED** [review]
i think your branch got messed up somehow - as it seems to roll back certain chances we made lately

**Comment by HaddadJoe** [comment]
Nice suggestion, I'm interested in this feature

@NoahGago are you still working on this? I can pick up where you left and clean that PR if you're not.

**Comment by NoahGago** [comment]
Nope I am not! You're free to pick up where I left off!

**Comment by xataxxx** [comment]
Totally agree that this should be exchange-agnostic - on KuCoin one might want to buy KCS, on FTX exchange you would want to buy FTT etc.

Besides running at Xh intervall maybe consider running at startup + at the end of SELL order finishing.
Basically re-check at startup + after each sell order closing check if we have enough fee coin left and buy more if needed.

Idea for future (way overkill for first implementation):
Would be awesome if we could calculate and check the full dataframe for the <exchange-coin>/<stake-currency> before buying so that the strategy could advise the bot if the BNB/KCS/FTT pair is having a bad moment and maybe skip buying if some indicators indicate that the price is very high temporarily etc. Then you could also "optimise" the buying of your fee currency when the BNB/KCS/FTT etc is currently selling relatively low.

**Comment by xmatthias** [comment]
Going to close this as it's abandoned, and there's still too much missing for this to be used as reasonable base.

**Pull request closed without merging**